### DISCUSSION TITLE
Azure CI brainstorming

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/llama.cpp#1985] [Ideas]
If we assume we had enough [MS Azure](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes-gpu) credits, how difficult would it be to make a `llama.cpp` / `whisper.cpp` / `ggml` CI in the cloud that runs on every commit and does some performance and perplexity benchmarks across a variety of hardware and models?

I don't have any experience in this, so would like to hear some opinions and maybe also see if there are people that would be interested in implementing such type of CI.

**Comment by ianscrivener** [comment]
@ggerganov,
A good idea and pretty easy/doable! I'd be happy to do this and submit a PR.

Let's iterate the solution on llama.cpp, then copy/paste the solution to whisper.cpp & ggml when it is "good enough"

TBH: I've been wanting to contribute to llama.cpp... but haven't had the matching tech skills until now.

**Comment by AlphaAtlas** [comment]
If you can secure one, Oracle's free Ampere instances would be a good candidate for continuous testing on ARM. Its 4 Ampere cores, 24GB, and enough ssd storage space for a few models.

**Comment by AlphaAtlas** [comment]
Also, one issue with VM cloud benchmarking is that performance is inconsistent.

If you do twenty runs of the same benchmark, the performance spread could be huge. Then you might spin up the exact same type of instance the next day, run twenty more runs of a benchmark, and then performance is very consistent.... but somehow slower than the previous sporadic run. I am not sure why this is the case, but I'm sure it has to do with other tenants stressing the CPU and its subsystems, variable silicon binning and such.

So if performance metrics are important, consider a small bare metal instance?

The Oracle VMs previously mentioned suffer from this, but since they are free and 24-7, you can probably get around it with a huge number of tests.

**Reply by ianscrivener** [reply]
Good point.
Yes, bare metal / dedicated VM is essential for permormance testing

**Comment by ianscrivener** [comment]
I've got cloud accounts with Google, Amazon & Microsoft Azure... a lot of free capacity with Azure & Google

Curiously, I've had problems signing up with Oracle (in Australia). Tried twice unsuccessfully, not sure if it my email address, credit card other?

AWS also has ARM machines - Inferentia & Trainium. IMO with the nvidia shortages Amazon's custom silicon will be getting a lot more more popular. Huggingface apparently use them for most (all?) training.

@ggerganov mentioned Azure - so I'll do a POC there. Assume he has some freebie on offer. I just signed up for USD$4500/year with Azure for a non-profit I'm helping out.

AWS has bare metal Apple M2s - (later on) I'd like to do performance testing there... give than Apple Silicon is a "first class citizen" for llama.cpp. They have a minimum 24hours usage purchase - so there are costs involved unless we get a ggml.ai AWS freebie deal.

**Reply by ianscrivener** [reply]
MacStadium have baremetal Apple M2s for around USD$120/month - https://www.macstadium.com/pricing

Perhaps a quick/tactical solution for perpexity, latency & resources benchmarking?

**Reply by ggerganov** [reply]
I should get access to "Azure Credits", but still not sure what this means exactly. Will follow-up when I have more info

**Comment by howard0su** [comment]
Be aware that you will get different VM every time you create a new VM. In order to make the perf result stable, you may have to keep one VM running.

**Reply by ianscrivener** [reply]
Yep! We probably should go bare metal for performance testing... and machines that we benchmark with over months & years.

**Reply by mechanicmuthu** [reply]
I'm a nobody in these topics, but I believe
1. bare metal
2. keeping VM running
3. repeatable benchmarking 
4. benchmarking using m2 dedicated servers, etc
are all taking us away from the intent here. The intent is to run llama reliably in ANY cloud instance. (with the benchmarking per run for info). Users will choose whether they can afford these things themselves. Thats my view.

**Reply by ianscrivener** [reply]
Agree @mechanicmuth.. 

I'm still waiting for @ggerganov to organise dedicated test machines... 

in the mean time I might code something slightly different... that does perplexity only

BTW: everybody is somebody!

**Reply by gjmulder** [reply]
:+1: for bare metal, unless you're willing to run multiple extended tests and report percentile summaries.

Since `llama.cpp` is primarily bottlenecked by memory I/O, running on _any_ shared virtualized environment means `llama.cpp`'s performance can be randomly throttled by memory I/O from other coscheduled VMs. AFAIK most if not all virtualization solutions do not provide any memory I/O throughput guarantees, unlike virtualized CPU and network throughput.

**Reply by ianscrivener** [reply]
@gjmulder, agree Gary.
IMO perplexity latency benchmarking is pretty much useless unless we have quality **bare metal** or **dedicated host**. 

@ggerganov said he can get _infinite_ Azure... so I was thinking small dedicated host with small Nvidia GPU. Have researched this in depth.

Even whole machine GPU cloud services like RunPod, Vultr et al offer a Docker container on a mostly (but not garranteed dedicated) machine. If it was dong backups or updates if would effectively be a "noisey neighbour" situations and ruin the perpexity latency benchmarks credibility.

AWS has Mac Mini baremetal, but you need to buy a whole day's worth of time. Again, budget is required.

Recently MS Azure refused GPUs to a not-for-profit I'm helping out who has USD3.5k of MS credit. So I'll believe free **bare metal** or **dedicated host** from GPU from Azure when I've spun it up and sshed in and tested it.

I'm reluctant to start coding a **_perplexity score + perpexity latency banchmark_** systems without test hardware locked in. Code would be mostly useless without dedicated test hardware.

Though I have been pondering a different approach.. perplexity scores only, post scores to HuggingFace Data or somewhere so anyone can run benchmark perplexity and anyone can Python/Jupyter graphs of llama.cpp perplexity results. Perhaps even a distributed job queue to fill in the benchmark data gaps.

**Comment by ianscrivener** [comment]
@ggerganov, 

OK, so I have had an initial review, checked out the "[perplexity page](https://github.com/ggerganov/llama.cpp/discussions/406)", reviewed the Github actions to build the releases, and test run `./perplexity` etc.

Some questions - please pls confirm/correct my assumptions

**Q1: Test Variables**
1. llama.cpp release id - *eg master-bd34cdd*
2. hardware - *eg Mac Mini 2022 M2 16Gb*
3. model - *eg /models/7B/ggml-model-q4_0.bin*
4. OS+GPU+Engine: _MacOS-M2-Metal | Windows-CUDA-OpenBLAS | Linux-CUDA-BLAS_ .. etc etc
5. ?? context size - _eg 512, 1024, 2048_  (unsure - pls confirm)     
6. ?? other command line options  (unsure - pls confirm)   

**Q2: Test Results Required**
1. perplexity score - _eg 4.2701_
2. perplexity test duration - _eg 3.74 hours, or 13,464seconds._. according to the `perplexity` process
3. ?? time according to `time` (unsure - pls confirm)   
4. ?? full `./perplexity`  process stdout  (unsure - pls confirm)   

**Q3: How will we present/display the results**
  1. web page / web dashboard
  2. Jupyter notebook
  3. rendered static graphs to be inserted in markdown READMEs
  4. all of the above (in time)

**Q4: Where/how to record test results?**
  1. database 
  2. Github as json, csv, other 
  3. all of the above 

**Q5: Test Hardware & OS**
  1. MacOS M2
  2. Intel + Nvidia GPU - Windows
  3. Intel + Nvidia GPU - Ubuntu
  4. AMD + AMD GPU - Windows
  5. AMD + AMD GPU - Ubuntu
  6. ARM Ampere (cloud) - Ubuntu
  7. Raspberry Pi (edge) - Raspbian
  8. Smaller Arm - eg Nuc (edge) - Ubuntu
  9. ?? other

**Reply by ggerganov** [reply]
So, I think we will aim for maximum hardware coverage - whatever is available in terms of OS, CPU, GPU, etc., we should have the tests running on it.

Regarding the model - we have to be careful. Either use OpenLLaMA, or something that we are confident will not cause us any trouble. Hosting `ggml-model-q4_0.bin` might be problematic due to LLaMA license. 

Perplexity runs are time consuming, so we might want to run partial runs (i.e. just a few iterations) to make sure results stayed the same as before. Maybe a way to run full perplexity runs from time to time, manually

But in any case, I think we should start with something basic, so that I can get a better idea of how this machinery works and know what's possible. Maybe start with a Mac Mini running some of the `llama.cpp` unit tests and figure out a way to present the results. Results should be something very simple. No databases. Just text, maybe HTML

**Reply by ianscrivener** [reply]
OK.
Yes, I was suprised to see ETA 3 hours when I first ran `./perplexity`

So a single model: OpenLLaMA.
Q1: Will we want to run multiple bit size/quantiztions of the model: FP16, q4_0 etc ?
Q2: would we run (in time) different sizes of the models: 3B, 7B, 13B?

BTW:
first thoughts on archicture & applications are;
- a completely separate project & git repo. 
- a tiny orchestation cloud VM that polls the Github API for releases.. so completely decoupled from llama.cpp codebase
- a basic vanilla node.js application with good enough documentation
- yaml config file(s)... much like Github Actions 
- (in time) specfic 'jobs' could be initiated via webhooks
- (in time) new yaml job files could be posted to an API

principals being;
1. opensource
2. lean, uncomplex
3. reusable
4. no vendor lockin/preferences

I did ponder using a cloud platform like Terraform Cloud.. but IMO the above solution approach has more velocity, openness and extensibility. I do like/resonate with the llama.cpp ethos of **lean, fast, working**.

**Reply by ghost** [reply]
Hi, I barely understand the technical details in this thread, but it appears there's mention of testing on various hardware.

I hope the hardware being tested also includes Android devices. For example, I use a Samsung S10+.

Here's some specifications:
uname -a
```
Linux localhost 4.14.190-23725627-abG975WVLS8IWD1 #2 SMP PREEMPT Mon Apr 10 18:16:39 KST 2023 aarch64 Android
```
lscpu:
```
Architecture:           aarch64
  CPU op-mode(s):       32-bit, 64-bit
  Byte Order:           Little Endian
CPU(s):                 8
  On-line CPU(s) list:  0-7
Vendor ID:              Qualcomm
  Model name:           Kryo-4XX-Silver
    Model:              14
    Thread(s) per core: 1
    Core(s) per socket: 4
    Socket(s):          1
    Stepping:           0xd
    CPU(s) scaling MHz: 100%
    CPU max MHz:        1785.6000
    CPU min MHz:        300.0000
    BogoMIPS:           38.40
    Flags:              fp asimd evtstrm aes pmull
                         sha1 sha2 crc32 atomics f
                        php asimdhp cpuid asimdrdm
                         lrcpc dcpop asimddp
  Model name:           Kryo-4XX-Gold
    Model:              14
    Thread(s) per core: 1
    Core(s) per socket: 2
    Socket(s):          2
    Stepping:           0xd
    CPU(s) scaling MHz: 83%
    CPU max MHz:        2841.6001
    CPU min MHz:        710.4000
    BogoMIPS:           38.40
    Flags:              fp asimd evtstrm aes pmull
                         sha1 sha2 crc32 atomics f
                        php asimdhp cpuid asimdrdm
                         lrcpc dcpop asimddp
```

GPU:
`
QUALCOMM Snapdragon(TM)
`

I think including testing for Android devices is necessary as it sucks when there's an incompatible release.

Thank you.

**Reply by SlyEcho** [reply]
> Yes, I was suprised to see ETA 3 hours when I first ran `./perplexity`

There has been some mention of using a truncated wiki.test.raw with [only the first 406 lines](https://github.com/ggerganov/llama.cpp/discussions/406).

**Reply by ianscrivener** [reply]
thanks @SlyEcho .. I'll post a question on that Discussion page

**Reply by ggerganov** [reply]
> Q1: Will we want to run multiple bit size/quantiztions of the model: FP14, q4_0 etc ?

Yes - ideally, run all quantization types. Could try to design the CI to have "frequent" runs of a single quantization type and "rare" (i.e. daily / nightly) runs on the rest to save time.

> Q2: would we run (in time) different sizes of the models: 3B, 7B, 13B?

Yes - probably even bigger models if we can apply the same "frequent" / "rare" strategy from above

**Reply by ianscrivener** [reply]
👍🏻 OK, I like the **_tiered approach to CI priorities_**

**Reply by ggerganov** [reply]
> a completely separate project & git repo

I made a CI team and repo in `ggml-org` - can invite anyone interested in collaborating on this project.
Let me know and I'll send an invite:

- repo: https://github.com/ggml-org/ci
- team: https://github.com/orgs/ggml-org/teams/ci

Alternatively - I'm also open to other suggestions for hosting / developing the framework

**Comment by gjmulder** [comment]
@ianscrivener, replying here as this about performance and not perplexity per se:

> my understanding (perhaps wrong) is that we are aiming to get
> (1) **perplexity score** - eg 8.2964 (per my test above)
> (2) **perplexity test runtime** - eg 360.09s (per my test above)
> 
> for a given a combination of
> (a) **model** - eg open-llama-3b-q4_0.bin
> (b) **hardware +GPU Library** - eg Mac M2 2022 - 8 CPU cores (4 performance, 4 efficiency) - 16Gb RAM + Metal
> (c) **test data** - eg wiki.test.raw.406
> (d) **context size** - eg 512, 1024, 2048

If you're looking for performance (2) your dependent variable is execution time. (a), (c), (d) should be kept constant and you should vary (b). (1) is irrelevant then, as for CI/CD it shouldn't change (at least say to `n` decimal places - e.g. you might find minor differences in math libraries across different (b)). You could however simply test that the final perplexity doesn't change as a simple functional regression verification.

My point was that you may want to judiciously choose (c) and your llama config params so you get timings for all of the 1 = sample, 2 = prompt and 3 = eval time and 1 + 2 + 3 = total compute time. Ignore the load time as that's a function on the model size and disk I/O of the VM the model is running on. Perplexity looks to only test load and prompt time. @ggerganov can confirm if just prompt time is sufficient for a performance test.

Furthermore, you want to ensure the runs are of sufficient duration that the potentially high level of variability between different successive virtualized instances will be somewhat averaged out. I'd use just 2 (or maybe 4) vCPUs (i.e. `-t 2`) to attempt to reduce variability between successive runs, as the more vCPUs you ask for the more chance of contention between your performance test VM and other co-hosted VMs. I'd look to configure the runtime of each test for at least 10 minutes, but around 60 minutes would be better if you have the compute. One way to determine how much test duration you can/should use is look at the approx. frequency of commited PRs (e.g. one per 6 hours) and ensure your performance test completes before the next PR is committed. I've seen cases where multiple rapid PR commits DoSed the CI/CD compute infrastructure. I'm not sure if Github Actions queues test runs if the previous run is still executing. 

Then, to deal with the inevitable variability of running on VMs when looking for non-functional performance regressions, I usually calculate an [exponentially smoothed time series model](https://www.statsmodels.org/0.9.0/generated/statsmodels.tsa.holtwinters.SimpleExpSmoothing.fit.html) of previous CI/CD total performance test run time results for each (b). You can then compare the last build for each (b) to this average and fail the build on performance if it is say 0.25 s.d. outside the predicted performance of the (b). I check both for upper and lower bounds as I've seen performance regressions where performance tests take 0 seconds to run :facepalm:

**Reply by ianscrivener** [reply]
excellent points, thanks @gjmulder !
your response gives me confidence that I'm starting to understand the problem space sufficiently

• there's been some discussion that we ideally will use **baremetal GPU machines** to eliminate the performance variability of shared compute. I'll likely dev/test on shared compute to start with... until we have budget or cloud freebies for baremetal.

• I'm intending to code a cloud vendor agnostic solution - so it could be run locally, in Github Codespaces, or on any cloud. Post the results to a simple API.

• @SlyEcho suggested the "truncated wiki.test.raw perplexity test" - specifically the first 406 lines of wiki.test.raw. **Ideally the llama.cpp dev community will agree on test data & test data size eventually** - so the test results can be sensibly compared over time

**Reply by gjmulder** [reply]
One approach is to dockerize everything using Github Actions and Docker Hub and then have your test harness poll Github for new Docker images. That way you easily can plug and play across different (virtualized) hardware for your CI/CD tests. 

Note that if you have no control over the virtualized hardware specs, you would need to explicitly set the CMake build flags as the dynamic CPU instruction support detection could say compile for AVX512 if you build env happens to have a recent CPU, or not AVX512 if the CPU is less modern.

I've proposed something similar for functionally testing downstream [llama-cpp-python precompiled wheels](https://github.com/abetlen/llama-cpp-python/issues/243), as half of the issues we're seeing are build related, partly due to the rapid upstream development of llama.cpp.

EDIT: https://github.com/abetlen/llama-cpp-python/issues/243, not https://github.com/abetlen/llama-cpp-python/issues/400

**Reply by ianscrivener** [reply]
A good idea... 
I have considered this approach... though we'd be dependent on the GPU support that Docker provides.😞 

I love Docker and it is one of my most used techs (spent all day doing Docker today). But...

Getting Docker to handle 6-9 month old nvidia GPUs is fine... but I reckon newer nvidia GPUs and non-nvidia GPUs would be a major PITA. And forget about AWS Inferentia and other ARM silicon. MacOS inside Docker with Metal GPU working well - I doubt we'll ever see that become a reality. Remeber that MacOS is a "first class citizen" for llama.cpp

"Inference at the edge" - yes Raspberry Pi has been mentioned for test hardware,

IMO baremetal is best!

**Reply by gjmulder** [reply]
Baremetal solves a lot functional and non-functional problems - if you have the budget to obtain it. 

I spent most of my performance testing professional life constantly needing to control for variables such as virtualized compute, variable network performance, etc. as I worked in large corporate environments where if you wanted physical tin you had to convince your manager to buy it and then hide it under one's desk :rofl:

**Reply by ianscrivener** [reply]
Agree. Baremetal is best for performance comparisons.
@ggerganov 's spec said "_enough_" of Azure credit.
Most baremetal GPU can be purchased by the minute at less than USD$1/hour.

**Reply by ggerganov** [reply]
I like the discussion!

> Perplexity looks to only test load and prompt time. @ggerganov can confirm if just prompt time is sufficient for a performance test.

We can start with just prompt time of `perplexity` tool. At some point we might want to look into `eval time` of `main` tool as well, but to avoid complicating things at the start, lets focus just on `perplexity` for now.

> @ggerganov 's spec said "enough" of Azure credit.

Yes - for now lets operate on the assumption that we will have "infinite" amount of Azure credits.

**Reply by ianscrivener** [reply]
OK.. "infinite" would probably be enough 👍 😀

**Comment by ianscrivener** [comment]
Some back of the _**ballpark guestimates & assumptions**_
```
a) Models        3	  Open Llama 3B, 7B, 13B
b) Context size  3	  512, 1024, 2048
c) Test data     1	  wiki.test.raw.406
d) Hardware      12       MacOS Metal, MacOS OpenBLAS, Nvidiax5, AMDx3, Inferentia, another ARM
                          = 108 tests
                          @ 5 mins per test (guess #1)
                          = 540 mins testing per release
                          = 9 hours testing per release 
                          @ $0.50/hour average (guess #2)
                          = $4.50 per release
```

**Reply by ianscrivener** [reply]
BTW: @ggerganov also suggested we start off with a single MacMini M2

**Reply by gjmulder** [reply]
You can probably skip the context size of 1024 and reduce your costs by 1/3 considering this BLAS-enabled smoke test:

![image](https://github.com/ggerganov/llama.cpp/assets/10833500/560ba3bf-cb04-4452-9c64-dddce71ee965)

**Reply by ianscrivener** [reply]
Copy that @gjmulder .. makes sense

**Reply by ianscrivener** [reply]
```
a) Models        3	  Open Llama 3B, 7B, 13B
b) Context size  2	  512, 2048
c) Test data     1	  wiki.test.raw.406
d) Hardware      12       MacOS Metal, MacOS OpenBLAS, Nvidiax5, AMDx3, Inferentia, another ARM
                          = 72 tests
                          @ 5 mins per test (guess #1)
                          = 360 mins testing per release
                          = 6 hours testing per release 
                          @ $0.50/hour average (guess #2)
                          = $3.00 per release
```

**Reply by gjmulder** [reply]
5 minutes average per run looks like a good guesstimate for your CuBLAS and similar accelerated runs. I'd factor 10x that time for your OpenBLAS runs.

Here's a [Gsheets matrix of perplexity runs](https://docs.google.com/spreadsheets/d/1UsbivogLMrQbBA-Fk0ESRGTrvCsknBUieSykfWn6D9Q/edit#gid=0) @TheBloke and I performed with a selection of FB llama-based models. The "perf" column (in minutes) was the estimated perplexity runtime as reported by `./perplexity`. The `efficiency = 1 / (perf * perplexity)` column was a metric I devised to compare the quality versus throughput trade-off for each result. There's numbers there for early BLAS enabled non-GPU accelerated runs at the bottom.

**Reply by ianscrivener** [reply]
brilliant! the spreadsheet help with my learnind curve. Thanks

**Reply by ianscrivener** [reply]
@gjmulder - question: what is the "perf" column on your spreadhseet please?

**Reply by gjmulder** [reply]
@ianscrivener the "perf" column (in minutes) was the estimated perplexity runtime as reported by `./perplexity`

**Reply by ianscrivener** [reply]
@gjmulder - copy that. Thanks

**Comment by ianscrivener** [comment]
@ggerganov - I just dropped you an email regarding CI 😀

**Comment by tpaviot** [comment]
Hi guys, I see this issue related to Azure and CI. FYI, I set up a build/test/deploy workflow on my free azure account, originally to be able to build/download quantized models directly on azure and save time/storage/memory on my laptop. It could also serve as a basis for CI.

https://github.com/tpaviot/llama.cpp-azure is a separate project so far, I just changed it to public a couple of hours ago (it used to be a private repo), it could be a dev branch as well.

**Reply by ianscrivener** [reply]
thanks for sharing @tpaviot 👍

**Reply by ggerganov** [reply]
Thanks and sorry for delay on this topic - will try to setup things soon. I have the access now to the Azure Credits - need to learn how to create the VMs and use them

**Reply by ianscrivener** [reply]
@ggerganov, 

I can setup things up if you give me Azure access... I'll email you again

BTW: usually you need to request access to GPU servers - they have safety limits set so people don't spends thousands accidently
It takes a day of so to get activated...

**NCasT4_v3-series** are a good **_shared_** GPU VM using T4.. to start playing around with

For proper latency benchmarking we need an Azure 'dedicated host'  - ie a whole server machine - no noisey neighbours. Likely you should talk to your MS rep about this.. it will be a special order on a freebie plan. Beware; it seems to be standard MS Redmond policy for "no GPUs on freebie plans"... but likely they'll carve out an exceptino for llama.cpp

(Non GPU) Dedicated Host info is here - https://azure.microsoft.com/en-au/pricing/details/virtual-machines/dedicated-host/.. from about USD$2800 per month

**Reply by ggerganov** [reply]
@ianscrivener and all

Thanks for the info and apologise for not responding to email

I went ahead and implemented a custom CI framework with a few Bash scripts:

- https://github.com/ggml-org/ci
- https://github.com/ggerganov/ggml/pull/393

For now I plan to spin-up a few VMs (x86 and Arm) and polish the CI in the `ggml` repository - add more tests, coverage, text generation. After I feel it is in a good state, will expand it to `llama.cpp` and `whisper.cpp` and look into renting GPU and bare-metal instances for performance and perplexity tests.

This framework is a bit of a hack, but it allows me to have control and good understanding of how it works. This way I can iterate more quickly on it.

**Reply by ianscrivener** [reply]
@ggerganov,
yes... sounds like the best idea. CI/CD needs to be tightly coupled with development and probably setup by the developers. Azure CI sounds good... but is very bloated. Good for big corporates - but not so much for a fast paced opensource project with a distrubuted team. Github Action, Bash and Docker are more suitable tools.

You already have perplexity bash scripts... so can get scores per release... but it is manual to record that. IMO what is missing is;
  1) **easily/automatically** gathering the perplexity test results (removing the manual processing)
  2) storing it centrally
  3) presenting this data 

That is the bit I have been playing with:
 (a) python code that wraps ./perplexty and generates a json file of the perplexity test run, sending the json to a web API
 (b) the API that takes the JSON and puts it in a data store (GoogleSheets for now)
 (c) easy viewing of scores (Google Sheets for now - per @gjmulder & @TheBloke)
  ... I'll send a PR for when it is ready
 
Latency benchmarking is a whole new beast - and needs dedicated hardware to be empirical and to be meaningful over time. 
 
👀 I love the architectural approach of [lzbench](https://github.com/inikep/lzbench).. put competing binaries into a wrapper processes and test them on the same hardware at the same time... so my fork of this idea is a **_"LLM inference engine head-to-head test engine"_**

**Comment by ianscrivener** [comment]
@tpaviot I checkout out your code... looks like a good start.  

Here's the bash code I use to setup and run a llama.cpp perplexity test on Docker container with CUDA support: https://gist.github.com/ianscrivener/71bde7a2bfc92e8d217900229d78df51

**Comment by ianscrivener** [comment]
In response to @ggerganov's call for perplexity and latency testing for llama.cpp I've coded **[llama.cpp perplexity scorecard](https://github.com/ianscrivener/llama-cpp-perplexity-scorecard)**... a helper project to run and gather `./perplexity` results, uploads JSON data to AWS S3 for review, analysis and charting.
